pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -149,7 +189,13 @@ func (p *ReaderPool) NewBinaryReader(ctx context.Context, logger log.Logger, bkt
 	}
 
 	if p.lazyReaderEnabled {
-		reader, err = NewLazyBinaryReader(ctx, readerFactory, logger, bkt, dir, id, p.metrics.lazyReader, p.onLazyReaderClosed)
+		lazyBinaryReader, lazyErr := NewLazyBinaryReader(ctx, readerFactory, logger, bkt, dir, id, p.metrics.lazyReader, p.onLazyReaderClosed)
+		if p.eagerLoadReaderEnabled && p.lazyLoadedHeadersSnapshot != nil {
+			if p.lazyLoadedHeadersSnapshot.IndexHeaderLastUsedTime[id] > 0 {
+				lazyBinaryReader.EagerLoad()
```

Comment:
lazyBinaryReader can be nil here if lazyErr is non-nill

Reply:
Added lazyErr must be nil check.

Reply:
Should we just return immediately if `lazyErr` is not `nil`?